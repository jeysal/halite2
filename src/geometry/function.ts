import cubicSpline from 'cubic-spline';

export const interpolate = (...knots: number[][]) => (point: number): number =>
  cubicSpline(point, knots.map(([x]) => x), knots.map(([_, y]) => y));
