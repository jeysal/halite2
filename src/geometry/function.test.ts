import { interpolate } from './function';

describe('interpolate', () => {
  test('calculates appropriate intermediate values', () => {
    const fn = interpolate([0, 2], [1, 3], [2, 1], [3, -10]);

    expect([0.1, 0.9, 1.1, 1.9, 2.1, 2.9].map(fn)).toMatchSnapshot();
  });

  test('hits the knots exactly', () => {
    const fn = interpolate([0, 2], [1, 3], [2, 1], [3, -10]);

    expect([0, 1, 2, 3].map(fn)).toEqual([2, 3, 1, -10]);
  });
});
