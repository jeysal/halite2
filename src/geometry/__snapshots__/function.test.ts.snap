// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interpolate calculates appropriate intermediate values 1`] = `
Array [
  2.1198,
  2.9342,
  3.052,
  1.5960000000000005,
  0.27619999999999917,
  -8.6822,
]
`;
